File: GamePad-Mobile/mobile/lib/core/network/performance_monitor.dart
/// Tracks send-side performance metrics over a rolling window.
class PerformanceMonitor {
  int _packetsSent = 0;
  int _bytesSent = 0;
  int _pongsReceived = 0;
  int _lastPingSeq = -1;
  int _lastPingSentMs = 0;
  int _lastRttMs = 0;
  int _minRttMs = 0;
  int _maxRttMs = 0;
  int _totalRttMs = 0;
  int _rttSamples = 0;

  // Rolling 1-second window
  final List<int> _sendTimestamps = [];
  int _lastCleanupMs = 0;

  /// Call this when sending a packet.
  void recordSend(int byteCount, int sequenceNumber) {
    _packetsSent++;
    _bytesSent += byteCount;
    final now = _now;
    _sendTimestamps.add(now);
    _cleanup(now);

    if (sequenceNumber >= 0) {
      _lastPingSeq = sequenceNumber;
      _lastPingSentMs = now;
    }
  }

  /// Call this when a PONG is received.
  void recordPong(int echoSeq) {
    _pongsReceived++;
    if (echoSeq == _lastPingSeq && _lastPingSentMs > 0) {
      final rtt = _now - _lastPingSentMs;
      _lastRttMs = rtt;
      _totalRttMs += rtt;
      _rttSamples++;

      if (_minRttMs == 0 || rtt < _minRttMs) _minRttMs = rtt;
      if (rtt > _maxRttMs) _maxRttMs = rtt;
    }
  }

  int get _now => DateTime.now().millisecondsSinceEpoch;

  void _cleanup(int now) {
    if (now - _lastCleanupMs < 250) return;
    _lastCleanupMs = now;
    _sendTimestamps.removeWhere((ts) => now - ts > 1000);
  }

  /// Returns a snapshot of current metrics.
  MetricsSnapshot getSnapshot() {
    _cleanup(_now);

    return MetricsSnapshot(
      packetsSent: _packetsSent,
      bytesSent: _bytesSent,
      pongsReceived: _pongsReceived,
      packetRate: _sendTimestamps.length,
      lastRttMs: _lastRttMs,
      minRttMs: _minRttMs,
      maxRttMs: _maxRttMs,
      avgRttMs: _rttSamples > 0 ? _totalRttMs ~/ _rttSamples : 0,
    );
  }

  void reset() {
    _packetsSent = 0;
    _bytesSent = 0;
    _pongsReceived = 0;
    _lastPingSeq = -1;
    _lastPingSentMs = 0;
    _lastRttMs = 0;
    _minRttMs = 0;
    _maxRttMs = 0;
    _totalRttMs = 0;
    _rttSamples = 0;
    _sendTimestamps.clear();
  }
}

/// Immutable performance data snapshot.
class MetricsSnapshot {
  final int packetsSent;
  final int bytesSent;
  final int pongsReceived;
  final int packetRate;
  final int lastRttMs;
  final int minRttMs;
  final int maxRttMs;
  final int avgRttMs;

  const MetricsSnapshot({
    this.packetsSent = 0,
    this.bytesSent = 0,
    this.pongsReceived = 0,
    this.packetRate = 0,
    this.lastRttMs = 0,
    this.minRttMs = 0,
    this.maxRttMs = 0,
    this.avgRttMs = 0,
  });

  @override
  String toString() =>
      'TX:$packetsSent | rate:$packetRate pkt/s | RTT:${lastRttMs}ms (avg:${avgRttMs}ms)';
}

File: GamePad-Mobile/pc-server/Core/PerformanceMonitor.cs
using System.Collections.Concurrent;

namespace GamePadServer.Core;

/// <summary>
/// Tracks real-time network performance metrics over a rolling window.
/// </summary>
public class PerformanceMonitor
{
    private long _packetsReceived;
    private long _packetsCrcError;
    private long _bytesReceived;
    private int _lastSeqNumber = -1;
    private long _packetsLost;

    // Rolling window (1 second) for rate calculation
    private readonly ConcurrentQueue<long> _packetTimestamps = new();
    private readonly ConcurrentQueue<(long ticks, int size)> _byteHistory = new();
    private long _lastCleanupTicks;

    private const long TicksPerSecond = 10_000_000;
    private const int RollingWindowMs = 1000;

    /// <summary>
    /// Reports a received packet to update all metrics.
    /// </summary>
    public void RecordPacket(int byteCount, int sequenceNumber)
    {
        Interlocked.Increment(ref _packetsReceived);
        Interlocked.Add(ref _bytesReceived, byteCount);

        var now = DateTime.UtcNow.Ticks;
        _packetTimestamps.Enqueue(now);
        _byteHistory.Enqueue((now, byteCount));

        // Detect packet loss from sequence number gaps
        if (_lastSeqNumber >= 0)
        {
            var expected = (_lastSeqNumber + 1) & 0xFFFF;
            if (sequenceNumber != expected)
            {
                var lost = (sequenceNumber - expected) & 0xFFFF;
                Interlocked.Add(ref _packetsLost, lost);
            }
        }
        _lastSeqNumber = sequenceNumber;

        CleanupOldEntries(now);
    }

    /// <summary>
    /// Records a CRC error.
    /// </summary>
    public void RecordCrcError()
    {
        Interlocked.Increment(ref _packetsCrcError);
    }

    private void CleanupOldEntries(long nowTicks)
    {
        if (nowTicks - _lastCleanupTicks < TicksPerSecond / 4)
            return;

        _lastCleanupTicks = nowTicks;
        var cutoff = nowTicks - RollingWindowMs * TimeSpan.TicksPerMillisecond;

        while (_packetTimestamps.TryPeek(out var ts) && ts < cutoff)
            _packetTimestamps.TryDequeue(out _);

        while (_byteHistory.TryPeek(out var entry) && entry.ticks < cutoff)
            _byteHistory.TryDequeue(out _);
    }

    /// <summary>
    /// Returns a snapshot of current metrics.
    /// </summary>
    public MetricsSnapshot GetSnapshot()
    {
        CleanupOldEntries(DateTime.UtcNow.Ticks);

        return new MetricsSnapshot
        {
            PacketsReceived = Interlocked.Read(ref _packetsReceived),
            PacketsCrcError = Interlocked.Read(ref _packetsCrcError),
            PacketsLost = Interlocked.Read(ref _packetsLost),
            BytesReceived = Interlocked.Read(ref _bytesReceived),
            PacketRate = _packetTimestamps.Count,
            Bitrate = _byteHistory.Sum(e => e.size) * 8 / RollingWindowMs * 1000,
        };
    }

    /// <summary>
    /// Resets all counters.
    /// </summary>
    public void Reset()
    {
        Interlocked.Exchange(ref _packetsReceived, 0);
        Interlocked.Exchange(ref _packetsCrcError, 0);
        Interlocked.Exchange(ref _packetsLost, 0);
        Interlocked.Exchange(ref _bytesReceived, 0);
        _lastSeqNumber = -1;
        _packetTimestamps.Clear();
        _byteHistory.Clear();
    }
}

/// <summary>
/// Immutable snapshot of performance metrics.
/// </summary>
public readonly struct MetricsSnapshot
{
    public long PacketsReceived { get; init; }
    public long PacketsCrcError { get; init; }
    public long PacketsLost { get; init; }
    public long BytesReceived { get; init; }
    public int PacketRate { get; init; }       // packets/s (rolling 1s)
    public long Bitrate { get; init; }          // bits/s   (rolling 1s)

    public double LossRate => PacketsReceived + PacketsLost > 0
        ? (double)PacketsLost / (PacketsReceived + PacketsLost) * 100
        : 0;

    public override string ToString() =>
        $"RX:{PacketsReceived} | " +
        $"rate:{PacketRate} pkt/s | " +
        $"bitrate:{Bitrate / 1000.0:F1} kbps | " +
        $"loss:{LossRate:F2}% | " +
        $"crc_err:{PacketsCrcError}";
}
